### ISSUE TITLE
DictionaryKeyPropertyAttribute not working

### ISSUE THREAD

**Issue by Kation** [dotnet/wpf#7562]
### Description

Class define with `DictionaryKeyPropertyAttribute` still show `All objects added to an IDictionary must have a Key attribute or some other type of key associated with them.` error while building project.

### Reproduction Steps

Define a class with `DictionaryKeyPropertyAttribute`.
```
[DictionaryKeyProperty("TargetType")]
public class MyStyle
{
    public MyStyle()
    {

    }

    public Type TargetType { get; set; }
}
```
Use it in a xaml file.
```
<Window x:Class="WpfApp1.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:WpfApp1"
        xmlns:my="clr-namespace:WpfLibrary1;assembly=WpfLibrary1"
        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800">
    <Window.Resources>
        <ResourceDictionary>
            <my:MyStyle TargetType="{x:Type Button}"/>
        </ResourceDictionary>
    </Window.Resources>
    <Grid>
    </Grid>
</Window>
```

### Expected behavior

Build successful

### Actual behavior

Build failed

**Comment by tyl4r** [comment]
### Here's a workaround
```xaml
<!-- Set x:Key to have the same value as TargetType -->
<my:MyStyle x:Key="{x:Type Button}" TargetType="{x:Type Button}"/>
```

**Comment by Kation** [comment]
@tyl4r I means there have to fix this issure.
It's about xaml intellisense issure.
It can read successful by xaml reader.

**Comment by Nintynuts** [comment]
I'm also wondering why this isn't working. I've checked the source to work out how it obtains the key from `DictionaryKeyPropertyAttribute` and can't see why it would work for built-in types but not for custom ones (but it's quite hard to follow, so I might have missed it).

[Here](https://github.com/dotnet/wpf/blob/d73dd1dbd908398cef7f50e57812e55b90ed3676/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/InfosetObjects/XamlObjectWriter.cs#L2586
) seems to be where it gets key from the property named by the attribute, but I can't see anything which restricts the types it works with:
(also [here](https://github.com/dotnet/wpf/blob/d73dd1dbd908398cef7f50e57812e55b90ed3676/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/XamlObjectReader.cs#L482), but doesn't seem to use the value for anything) which throws `SR.MissingKey` which I can't seem to find.

But there are [two](https://github.com/dotnet/wpf/blob/d73dd1dbd908398cef7f50e57812e55b90ed3676/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Markup/BamlRecordWriter.cs#L1577) [places](https://github.com/dotnet/wpf/blob/d73dd1dbd908398cef7f50e57812e55b90ed3676/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Markup/BamlRecordReader.cs#L4660) where the MC3022 exception `SR.ParserNoDictionaryKey` is thrown.

**Comment by Nintynuts** [comment]
Related: https://github.com/dotnet/wpf/issues/2404
This really should be fixed (and back-ported to previous .NET versions).
Not only does it prevent extensibility but it means `ItemContainerTemplate` (for dynamic `MenuItems`) doesn't work properly either.

I have a strong suspicion that the reason this is broken is because the attribute is inherited
https://github.com/dotnet/wpf/blob/cdd7f1100892ac9677d36c273b81a8178d6cb2b3/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Windows/Markup/DictionaryKeyPropertyAttribute.cs#L8

So when you derive a class and add the attribute it's ignored because the inherited one is found first and the Reflector always uses the attribute at index 0, rather than the last index (which will be the most derived).

https://github.com/dotnet/wpf/blob/cdd7f1100892ac9677d36c273b81a8178d6cb2b3/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/Schema/Reflector.cs#L69

https://github.com/dotnet/wpf/blob/cdd7f1100892ac9677d36c273b81a8178d6cb2b3/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/Schema/Reflector.cs#L85-L88

This also explains why the `ItemContainerTemplate` doesn't work when you set `DataType`, because doing so replaces the `x:Key` value with a `DataTemplateKey` (you have to set `x:Key` and `d:DataContext` instead).
